service/storage_service.cc

```diff
@@ -2544,15 +2552,83 @@ class topology_coordinator {
+            // if coordinator allows for handling of tablet migrations.
+            _topo_sm.event.signal();
+        })
         , _ring_delay(ring_delay)
     {}
 
     future<> run();
 };
 
+future<locator::load_stats> topology_coordinator::refresh_table_load_stats() {
+    locator::load_stats stats;
+
+    auto tm = get_token_metadata_ptr();
+    auto& topology = tm->get_topology();
+
+    for (auto& [dc, endpoints] : topology.get_datacenter_endpoints()) {
+        locator::load_stats dc_stats;
+        slogger.info("raft topology: Refreshing table load stats for DC {} that has {} endpoints", dc, endpoints.size());
+        for (auto& endpoint : endpoints) {
+            auto* node = topology.find_node(endpoint);
+            if (!node) {
+                // TODO: that shouldn't happen, perhaps abort instead.
+                continue;
+            }
+            auto dst = node->host_id();
+
+            auto node_stats = co_await ser::storage_service_rpc_verbs::send_table_load_stats(&_messaging,
+                netw::msg_addr(id2ip(dst)),
+                _as,
+                raft::server_id(dst.uuid()));
+
+            dc_stats += node_stats;
+        }
+
+        for (auto& [table_id, table_stats] : dc_stats.tables) {
+            co_await coroutine::maybe_yield();
+
+            auto& t = _db.find_column_family(table_id);
+            auto& rs = t.get_effective_replication_map()->get_replication_strategy();
+            if (!rs.uses_tablets()) {
+                continue;
+            }
+            const auto* nts_ptr = dynamic_cast<const locator::network_topology_strategy*>(&rs);
+            if (!nts_ptr) {
+                // TODO: abort
+                continue;
+            }
+
+            auto rf_for_this_dc = nts_ptr->get_replication_factor(dc);
+            if (rf_for_this_dc <= 0) {
+                continue;
+            }
+            // Takes into account the RF of each DC, so stats will still make sense when aggregated
+            // from multiple DCs. This allows tablet allocator to compute an average tablet size
+            // by dividing total size in bytes by tablet count.
+            table_stats.size_in_bytes /= rf_for_this_dc;
+            slogger.info("raft topology: Refreshed table load stats for DC {}, table={}, RF={}, size_in_bytes={}, split_ready_phase={}",
+                          dc, table_id, rf_for_this_dc, table_stats.size_in_bytes, table_stats.split_ready_phase);
+        }
+
+        stats += dc_stats;
+    }
+
+    co_return std::move(stats);
+}
+
 future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard guard) {
     slogger.debug("raft topology: Evaluating tablet balance");
 
+    std::optional<locator::load_stats> load_stats;
+    if (lowres_clock::now() - _last_load_stats_refresh > load_stats_refresh_interval) {
+        load_stats = co_await refresh_table_load_stats();
+        _last_load_stats_refresh = lowres_clock::now();
+    }
```

Comment:
Like here. We don't want to stop migration if we weren't able to refresh statistics.

Reply:
Right, we need to make stats refresher independent and trigger migration path on completion.